File: dotnet/picocontainer/src/Defaults/AbstractComponentAdapter.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using PicoContainer;

namespace PicoContainer.Defaults
{
	/// <summary>
	/// Base for implementing ComponentAdapters
	/// </summary>
	[Serializable]
	public abstract class AbstractComponentAdapter : IComponentAdapter
	{
		private readonly object componentKey;
		private readonly Type componentImplementation;
		private IPicoContainer container;


		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="componentKey">The component's key</param>
		/// <param name="componentImplementation">The component implementing type</param>
		protected AbstractComponentAdapter(object componentKey, Type componentImplementation)
		{
			if (componentImplementation == null)
			{
				throw new NullReferenceException("componentImplementation");
			}
			this.componentKey = componentKey;
			this.componentImplementation = componentImplementation;
			CheckTypeCompatibility();
			CheckConcrete();
		}

		/// <summary>
		/// The key of the component
		/// </summary>
		public object ComponentKey
		{
			get
			{
				if (componentKey == null)
				{
					throw new NullReferenceException("componentKey");
				}
				return componentKey;
			}
		}

		/// <summary>
		/// The component implementing type
		/// </summary>
		public Type ComponentImplementation
		{
			get { return componentImplementation; }
		}

		private void CheckTypeCompatibility()
		{
			Type componentType = componentKey as Type;
			if (componentType != null)
			{
				if (!componentType.IsAssignableFrom(componentImplementation))
				{
					throw new AssignabilityRegistrationException(componentType, componentImplementation);
				}
			}
		}

		private void CheckConcrete()
		{
			if (componentImplementation.IsInterface || componentImplementation.IsAbstract)
			{
				throw new NotConcreteRegistrationException(componentImplementation);
			}
		}

		/// <summary>
		/// The string representation of the ComponentAdapter
		/// </summary>
		/// <returns>The name</returns>
		public override string ToString()
		{
			return this.GetType().Name + "[" + ComponentKey + "]";
		}

		/// <summary>
		///  Property containing the container in which this instance is registered, called by the container upon registration
		/// </summary>
		public IPicoContainer Container
		{
			get { return container; }

			set { this.container = value; }
		}

		/// <summary>
		/// Gets the component instance. This method will usually create
		/// a new instance for each call.
		/// </summary>
		/// <remarks>
		/// Not all ComponentAdapters return a new instance for each call an example is the <see cref="PicoContainer.Defaults.CachingComponentAdapter"/>.<BR/>
		/// </remarks>
		/// <returns>a component instance</returns>
		/// <exception cref="PicoContainer.PicoInitializationException">if the component could not be instantiated.</exception>    
		public abstract object ComponentInstance { get; }

		public abstract object GetComponentInstance(IPicoContainer container);

		/// <summary>
		/// Verify that all dependencies for this adapter can be satisifed.
		/// </summary>
		/// <exception cref="PicoContainer.PicoIntrospectionException">if the verification failed</exception>
		public abstract void Verify(IPicoContainer container);
	}
}

File: java/sandbox/baby-steps/pico2/container/src/java/org/picocontainer/adapters/AbstractComponentAdapter.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Original code by                                                          *
 *****************************************************************************/
package org.picocontainer.adapters;

import org.picocontainer.ComponentMonitor;
import org.picocontainer.PicoVisitor;
import org.picocontainer.adapters.MonitoringComponentAdapter;
import org.picocontainer.defaults.AssignabilityRegistrationException;
import org.picocontainer.defaults.DelegatingComponentMonitor;

/**
 * Base class for a ComponentAdapter with general functionality.
 * This implementation provides basic checks for a healthy implementation of a ComponentAdapter.
 * It does not allow to use <code>null</code> for the component key or the implementation, 
 * ensures that the implementation is a concrete class and that the key is assignable from the 
 * implementation if the key represents a type.   
 *  
 * @author Paul Hammant
 * @author Aslak Helles&oslash;y
 * @author Jon Tirs&eacute;n
 * @version $Revision$
 * @since 1.0
 */
public abstract class AbstractComponentAdapter extends MonitoringComponentAdapter {
    private Object componentKey;
    private Class componentImplementation;

    /**
     * Constructs a new ComponentAdapter for the given key and implementation. 
     * @param componentKey the search key for this implementation
     * @param componentImplementation the concrete implementation
     * @throws org.picocontainer.defaults.AssignabilityRegistrationException if the key is a type and the implementation cannot be assigned to.
     */
    protected AbstractComponentAdapter(Object componentKey, Class componentImplementation) throws AssignabilityRegistrationException {
        this(componentKey, componentImplementation, new DelegatingComponentMonitor());
    }

    /**
     * Constructs a new ComponentAdapter for the given key and implementation. 
     * @param componentKey the search key for this implementation
     * @param componentImplementation the concrete implementation
     * @param monitor the component monitor used by this ComponentAdapter
     * @throws AssignabilityRegistrationException if the key is a type and the implementation cannot be assigned to.
     */
    protected AbstractComponentAdapter(Object componentKey, Class componentImplementation, ComponentMonitor monitor) throws AssignabilityRegistrationException {
        super(monitor);
        if (componentImplementation == null) {
            throw new NullPointerException("componentImplementation");
        }
        this.componentKey = componentKey;
        this.componentImplementation = componentImplementation;
        checkTypeCompatibility();
    }

    /**
     * {@inheritDoc}
     * @see org.picocontainer.ComponentAdapter#getComponentKey()
     */
    public Object getComponentKey() {
        if (componentKey == null) {
            throw new NullPointerException("componentKey");
        }
        return componentKey;
    }

    /**
     * {@inheritDoc}
     * @see org.picocontainer.ComponentAdapter#getComponentImplementation()
     */
    public Class getComponentImplementation() {
        return componentImplementation;
    }

    protected void checkTypeCompatibility() throws AssignabilityRegistrationException {
        if (componentKey instanceof Class) {
            Class componentType = (Class) componentKey;
            if (!componentType.isAssignableFrom(componentImplementation)) {
                throw new AssignabilityRegistrationException(componentType, componentImplementation);
            }
        }
    }

    /**
     * @return Returns the ComponentAdapter's class name and the component's key.
     * @see java.lang.Object#toString()
     */
    public String toString() {
        return getClass().getName() + "[" + getComponentKey() + "]";
    }

    public void accept(PicoVisitor visitor) {
        visitor.visitComponentAdapter(this);
    }
    
}
